### ISSUE TITLE
libosmium 2.5.3 fails to build on various architectures due to test failures

### ISSUE THREAD

**Issue by sebastic** [osmcode/libosmium#135]
The uploaded of the libosmium (2.5.3-1~exp1) Debian package failed to build on [armel](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=armel&ver=2.5.3-1%7Eexp1&stamp=1447886797), [armhf](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=armhf&ver=2.5.3-1%7Eexp1&stamp=1447887081), [i386](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=i386&ver=2.5.3-1%7Eexp1&stamp=1447886250), [mips](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=mips&ver=2.5.3-1%7Eexp1&stamp=1447905249), [mipsel](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=mipsel&ver=2.5.3-1%7Eexp1&stamp=1447888541), [powerpc](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=powerpc&ver=2.5.3-1%7Eexp1&stamp=1447886376), [hppa](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=hppa&ver=2.5.3-1%7Eexp1&stamp=1447908526), [hurd-i386](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=hurd-i386&ver=2.5.3-1%7Eexp1&stamp=1447908793) & [kfreebsd-i386](https://buildd.debian.org/status/fetch.php?pkg=libosmium&arch=kfreebsd-i386&ver=2.5.3-1%7Eexp1&stamp=1447887675), this subsequently prevents the build of osmium-tool & pyosmium on these architectures.

The armel build failure can be ignored, because libosmium has never built successfully on that architecture.

The cause of the other build failures are the following two tests as taken from the armhf buildlog linked above:
```
test 12
      Start 12: basic_test_timestamp

12: Test command: /«PKGBUILDDIR»/build/test/basic_test_timestamp
12: Environment variables: 
12:  OSMIUM_TEST_DATA_DIR=/«PKGBUILDDIR»/test
12: Test timeout computed to be: 9.99988e+06
12: 
12: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12: basic_test_timestamp is a Catch v1.2.1 host application.
12: Run with -? for options
12: 
12: -------------------------------------------------------------------------------
12: Timestamp
12:   can be compared
12: -------------------------------------------------------------------------------
12: /«PKGBUILDDIR»/test/t/basic/test_timestamp.cpp:7
12: ...............................................................................
12: 
12: /«PKGBUILDDIR»/test/t/basic/test_timestamp.cpp:66: FAILED:
12:   REQUIRE( t1 < osmium::end_of_time() )
12: with expansion:
12:   1970-01-01T00:00:10Z
12:   <
12:   1969-12-31T23:59:59Z
12: 
12: ===============================================================================
12: test cases:  1 |  0 passed | 1 failed
12: assertions: 21 | 20 passed | 1 failed
12: 
12/52 Test #12: basic_test_timestamp .....................***Failed    0.02 sec
...
test 50
      Start 50: util_test_minmax

50: Test command: /«PKGBUILDDIR»/build/test/util_test_minmax
50: Environment variables: 
50:  OSMIUM_TEST_DATA_DIR=/«PKGBUILDDIR»/test
50: Test timeout computed to be: 9.99988e+06
50: 
50: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
50: util_test_minmax is a Catch v1.2.1 host application.
50: Run with -? for options
50: 
50: -------------------------------------------------------------------------------
50: minmax timestamp
50:   min
50: -------------------------------------------------------------------------------
50: /«PKGBUILDDIR»/test/t/util/test_minmax.cpp:38
50: ...............................................................................
50: 
50: /«PKGBUILDDIR»/test/t/util/test_minmax.cpp:44: FAILED:
50:   REQUIRE( x().to_iso() == "2010-01-01T00:00:00Z" )
50: with expansion:
50:   "1969-12-31T23:59:59Z"
50:   ==
50:   "2010-01-01T00:00:00Z"
50: 
50: ===============================================================================
50: test cases: 2 | 1 passed | 1 failed
50: assertions: 9 | 8 passed | 1 failed
50: 
50/52 Test #50: util_test_minmax .........................***Failed    0.01 sec
```

**Comment by joto** [comment]
I hope this will fix the problem. I'll release a new version soon.

**Comment by sebastic** [comment]
I've tested 1950853 as a patch for the Debian package, but these two tests still fail on i386 and I presume on the other 32bit architectures too.

**Comment by tomhughes** [comment]
Agreed - still failing for i386 on Fedora as well.

**Comment by tomhughes** [comment]
I'm also seeing the multipolygon test failing on i686, but not on arm:

* x86_64 is ok (https://kojipkgs.fedoraproject.org//work/tasks/9672/11919672/build.log)
* i686 is failing (https://kojipkgs.fedoraproject.org//work/tasks/9420/11919420/build.log)
* armv7hl is ok (https://kojipkgs.fedoraproject.org//work/tasks/9781/11919781/build.log)

The actual failure in this case is:

```
56: Test command: /usr/bin/cmake "-D" "OSM_TESTDATA=/builddir/build/BUILD/libosmium-2.5.3/osm-testdata" "-D" "RUBY=/usr/bin/ruby" "-P" "/builddir/build/BUILD/libosmium-2.5.3/libosmium/test/data-tests/run-testdata-multipolygon.cmake"
56: Test timeout computed to be: 9.99988e+06
56: 
56: 
56: 700...OK   701...OK   702...OK   703...OK   704...OK   705...OK   706...OK   707...OK   708...OK   709...OK   
56: 710...OK   711...OK   
56: 720...OK   721...OK   722...OK   723...OK   724...OK   725...OK   726...OK   727...OK   728...OK   729...OK   
56: 730...OK   731...OK   732...OK   733...OK   
56: 740...OK   741...OK   742...OK   743...ERR  744...OK   745...OK   746...OK   747...OK   748...OK   749...OK   
56: 750...OK   751...OK   752...OK   753...OK   754...OK   755...OK   756...OK   757...OK   758...OK   759...OK   
56: 760...OK   761...OK   762...OK   763...OK   764...OK   765...OK   766...OK   767...OK   
56: 770...OK   771...ERR  772...OK   773...ERR  
56: 780...OK   781...OK   782...OK   
56: 
56: 900...OK   901...OK   902...OK   903...OK   904...OK   905...OK   
56: 910...OK   911...OK   912...OK   913...OK   
56: 920...OK   921...OK   922...OK   923...OK   924...OK   925...OK   926...OK   927...OK   
56: 930...OK   931...OK   
56: 940...OK   
56: 950...OK   
56: CMake Error at /builddir/build/BUILD/libosmium-2.5.3/libosmium/test/data-tests/run-testdata-multipolygon.cmake:42 (message):
56:   Error running compare-areas command
56: 
56: 
56/56 Test #56: testdata-multipolygon ....................***Failed   26.18 sec
```

**Comment by joto** [comment]
I believe both problems have been fixed in the current master version. @sebastic and @tomhughes: could you check?

**Comment by tomhughes** [comment]
Yes that seems to have fixed everything.

**Comment by sebastic** [comment]
With 39fe3a5b5bf5ac2cebb4048f0c7aecd3a9d3c8be as and additional patch all the tests succeed on i386 & amd64 at least.

**Comment by sabas** [comment]
From master I get
```
[  3%] Building CXX object test/CMakeFiles/area_test_area_id.dir/t/area/test_area_id.cpp.o
In file included from /home/sabas/libosmium/include/osmium/osm/node_ref_list.hpp:42:0,
                 from /home/sabas/libosmium/include/osmium/osm/area.hpp:45,
                 from /home/sabas/libosmium/test/t/area/test_area_id.cpp:3:
/home/sabas/libosmium/include/osmium/osm/node_ref.hpp: In member function ‘constexpr double osmium::NodeRef::lon() const’:
/home/sabas/libosmium/include/osmium/osm/node_ref.hpp:96:35: error: call to non-constexpr function ‘double osmium::Location::lon() const’
             return m_location.lon();
                                   ^
/home/sabas/libosmium/include/osmium/osm/node_ref.hpp: In member function ‘constexpr double osmium::NodeRef::lat() const’:
/home/sabas/libosmium/include/osmium/osm/node_ref.hpp:105:35: error: call to non-constexpr function ‘double osmium::Location::lat() const’
             return m_location.lat();
                                   ^
make[2]: *** [test/CMakeFiles/area_test_area_id.dir/t/area/test_area_id.cpp.o] Error 1
make[1]: *** [test/CMakeFiles/area_test_area_id.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by joto** [comment]
@sabas: Thanks for the report. Fixed in 9b0602a. Next time please open a new issue instead of tacking the report onto an unrelated issue.

**Comment by sabas** [comment]
I'm sorry, I read test and thought it was related. Thanks!

**Comment by tomhughes** [comment]
I think this is all resolved and be closed now? Certainly 2.5.4 builds cleanly on all my platforms without any patches.

**Comment by joto** [comment]
Thanks, Tom.
